src/Metrics/Reporter.h

```diff
@@ -0,0 +1,68 @@
+// SPDX-License-Identifier: BSD-2-Clause
+//
+// This code has been produced by the European Spallation Source
+// and its partner institutes under the BSD 2 Clause License.
+//
+// See LICENSE.md at the top level for license information.
+//
+// Screaming Udder!                              https://esss.se
+
+#pragma once
+
+#include "InternalMetric.h"
+#include "Sink.h"
+#include <asio.hpp>
+#include <map>
+#include <memory>
+#include <thread>
+
+namespace Metrics {
+
+class Reporter {
+public:
+  Reporter(std::unique_ptr<Sink> MetricSink, std::chrono::milliseconds Interval)
+      : MetricSink(std::move(MetricSink)), IO(), Period(Interval),
+        AsioTimer(IO, Period){};
+
+  void reportMetrics() {
+    IO.post([&]() {
+      for (auto &MetricNameValue : MetricsToReportOn) {
+        MetricSink->reportMetric(MetricNameValue.second);
+      }
+    });
+  }
+
+  void addMetric(Metric &NewMetric, std::string const &NewName) {
+    IO.post([&]() {
+      MetricsToReportOn.emplace(NewName, InternalMetric(NewMetric, NewName));
+    });
+  }
+
+  void tryRemoveMetric(std::string const &MetricName) {
+    IO.post([&]() { MetricsToReportOn.erase(MetricName); });
```

Comment:
For removing a metric, this has to be a blocking call as the metrics thread might access the counter after it has been deallocated by the `Metric` instance. I have successfully used code similar to the following:
```
void tryRemoveMetric(...) {
  auto WorkDone = std::make_shared<std::promise<void>>();
  auto WorkDoneFuture = WorkDone->get_future();
  IO.post([&, WorkDone{std::move(WorkDone)} ]() {
    MetricsToReportOn.erase(MetricName);
    WorkDone->set_value();
  });
  WorkDoneFuture.wait();
}
```